File: migrate/dotnet/src/com/tacitknowledge/util/migration/ADO/DataSourceMigrationContext.cs
/* Copyright 2005 Tacit Knowledge LLC
*
* Licensed under the Tacit Knowledge Open License, Version 1.0 (the "License");
* you may not use this file except in compliance with the License. You may
* obtain a copy of the License at http://www.tacitknowledge.com/licenses-1.0.
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
#region Imports
using System;
using AutopatchNET.src.com.tacitknowledge.util.migration;
using com.tacitknowledge.util.migration;
using MigrationException = com.tacitknowledge.util.migration.MigrationException;
#endregion
namespace com.tacitknowledge.util.migration.ado
{
	
	/// <summary> Provides ADO resources to migration tasks.
	/// 
	/// </summary>
	/// <author>   Scott Askew (scott@tacitknowledge.com)
	/// </author>
	public class DataSourceMigrationContext : ADOMigrationContext
    {

        #region Members
        /// <summary> The database connection to use</summary>
        //UPGRADE_NOTE: There are other database providers or managers under System.Data namespace which can be used optionally to better fit the application requirements. "ms-help://MS.VSCC.v80/dv_commoner/local/redirect.htm?index='!DefaultContextWindowIndex'&keyword='jlca1208'"
        private System.Data.Common.DbConnection connection = null;

        /// <summary>Will provide the DBConnection required to talk to the data store</summary>
        
        private MigrationDataSource dataSource = null;

        /// <summary> The name of the system being patched</summary>
        private String systemName = null;

        /// <summary> The name of the system being patched</summary>
        private DatabaseType databaseType = null;

        private static Int32 MAX_SYSTEMNAME_LENGTH = 30;

        #endregion

        #region Methods
        /// <summary> Returns the database connection to use
		/// 
		/// </summary>
		/// <returns> the database connection to use
		/// </returns>
		/// <throws>  SQLException if an unexpected error occurs </throws>
		
		public System.Data.Common.DbConnection Connection
		{
			get
			{
				if ((connection == null) || (connection.State == System.Data.ConnectionState.Closed))
				{
					
					/*
                     * Obtain a connection object for talking to the Data store
                     */
                    dataSource = new MigrationDataSource();
                    connection = dataSource.getConnection();
                    return connection;
				}
				return connection;
			}
			
		}
		
		
		/// <seealso cref="ADOMigrationContext.Commit()">
		/// </seealso>
		public virtual void  Commit()
		{
			try
			{
				SupportClass.TransactionManager.manager.Commit(Connection);
			}
			catch (System.Data.OleDb.OleDbException e)
			{
				throw new MigrationException("Error committing SQL transaction", e);
			}
		}
		
		/// <seealso cref="ADOMigrationContext.Rollback()">
		/// </seealso>
		public virtual void  Rollback()
		{
			try
			{
				SupportClass.TransactionManager.manager.RollBack(Connection);
			}
			catch (System.Data.OleDb.OleDbException e)
			{
				throw new MigrationException("Could not rollback SQL transaction", e);
			}
		}
		
		/// <summary> Returns the type of database being patched.
		/// 
		/// </summary>
		/// <returns> the type of database being patched
		/// </returns>
		public virtual DatabaseType getDatabaseType()
		{
			return databaseType;
		}
		
		/// <summary> Returns the type of database being patched.
		/// 
		/// </summary>
		/// <param name="type">the type of database being patched
		/// </param>
		public virtual void  setDatabaseType(DatabaseType type)
		{
			this.databaseType = type;
		}
		
		/// <returns> Returns the systemName.
		/// </returns>
		public virtual System.String getSystemName()
		{
			return systemName;
		}
		
		/// <summary> Sets the system name.
		/// 
		/// </summary>
		/// <param name="name">the name of the system to patch
		/// </param>
		public virtual void  setSystemName(System.String name)
		{
			if (name == null)
			{
				throw new System.ArgumentException("systemName cannot be null");
			}
            if (name.Length > MAX_SYSTEMNAME_LENGTH)
			{
                throw new System.ArgumentException("systemName cannot be longer than " + MAX_SYSTEMNAME_LENGTH + " characters");
			}
			this.systemName = name;
        }
        #endregion
    }
}

File: migrate/src/java/com/tacitknowledge/util/migration/jdbc/DataSourceMigrationContext.java
/* 
 * Copyright 2007 Tacit Knowledge LLC
 *
 * Licensed under the Tacit Knowledge Open License, Version 1.0 (the "License");
 * you may not use this file except in compliance with the License. You may
 * obtain a copy of the License at http://www.tacitknowledge.com/licenses-1.0.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.tacitknowledge.util.migration.jdbc;

import java.sql.Connection;
import java.sql.SQLException;

import javax.sql.DataSource;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;

import com.tacitknowledge.util.migration.MigrationException;

/**
 * Provides JDBC resources to migration tasks.
 *
 * @author  Scott Askew (scott@tacitknowledge.com)
 */
public class DataSourceMigrationContext implements JdbcMigrationContext
{
    /** Class logger */
    private static Log log = LogFactory.getLog(JdbcMigrationContext.class);

    /** The database connection to use */
    private Connection connection = null;

    /** The DataSource to use */
    private DataSource dataSource = null;

    /** The name of the system being patched */
    private String systemName = null;

    /** The name of the system being patched */
    private DatabaseType databaseType = null;

    /**
     * Returns the database connection to use
     *
     * @return the database connection to use
     * @throws SQLException if an unexpected error occurs
     */
    public Connection getConnection() throws SQLException
    {
        if ((connection == null) || connection.isClosed())
        {
            DataSource ds = getDataSource();
            if (ds != null)
            {
                connection = ds.getConnection();
            }
            else
            {
                throw new SQLException("Datasource is null");
            }
        }
        return connection;
    }

    /** {@inheritDoc} */
    public void commit() throws MigrationException
    {
        try
        {
            if (getConnection().getAutoCommit())
            {
                log.warn("Attempt to call commit on connection with autoCommit engaged - breaks in JBoss");
                return;
            }
            getConnection().commit();
        }
        catch (SQLException e)
        {
            throw new MigrationException("Error committing SQL transaction", e);
        }
    }

    /** {@inheritDoc} */
    public void rollback() throws MigrationException
    {
        try
        {
            if (getConnection().getAutoCommit())
            {
                log.warn("Attempt to call rollback on connection with autoCommit engaged - breaks in JBoss");
                return;
            }
            getConnection().rollback();
        }
        catch (SQLException e)
        {
            throw new MigrationException("Could not rollback SQL transaction", e);
        }
    }

    /**
     * Returns the type of database being patched.
     *
     * @return the type of database being patched
     */
    public DatabaseType getDatabaseType()
    {
        return databaseType;
    }

    /**
     * Returns the type of database being patched.
     *
     * @param type the type of database being patched
     */
    public void setDatabaseType(DatabaseType type)
    {
        this.databaseType = type;
    }

    /**
     * @return Returns the dataSource.
     */
    public DataSource getDataSource()
    {
        return dataSource;
    }

    /**
     * @param dataSource The dataSource to set.
     */
    public void setDataSource(DataSource dataSource)
    {
        this.dataSource = dataSource;
    }

    /**
     * @return Returns the systemName.
     */
    public String getSystemName()
    {
        return systemName;
    }

    /**
     * Sets the system name.
     *
     * @param name the name of the system to patch
     */
    public void setSystemName(String name)
    {
        if (name == null)
        {
            throw new IllegalArgumentException("systemName cannot be null");
        }
        if (name.length() > MAX_SYSTEMNAME_LENGTH)
        {
            throw new IllegalArgumentException("systemName cannot be longer than "
                + MAX_SYSTEMNAME_LENGTH + " characters");
        }
        this.systemName = name;
    }
    
    /**
     * Useful for debugging
     * 
     * @return String with state information
     */
    public String toString()
    {
        return 
            "DataSourceMigrationContext[" 
            + getDatabaseType() + "/"
            + getSystemName() + "/"
            + getDataSource() + "]";
    }
}
